hashes/README.md

```diff
@@ -30,3 +30,90 @@ Alternatively add symlinks in your `.git/hooks` directory to any of the githooks
 
 We use a custom Rust compiler configuration conditional to guard the bench mark code. To run the
 bench marks use: `RUSTFLAGS='--cfg=bench' cargo +nightly bench`.
+
+## Algorithms
+
+We currently support the following hash algorithms
+
+### [SHA-1](https://en.wikipedia.org/wiki/SHA-1)
+
+Since 2005, SHA-1 has not been considered secure against well-funded opponents
+
+Module: `sha1`
+Digest: 160 bit (20 byte)
+
+### [SHA-2](https://en.wikipedia.org/wiki/SHA-2)
+
+Supported functions from the SHA-2 family:
+
+- SHA-256
+
+Module: `sha256`
+Digest: 256 bit (32 byte)
+
+- SHA-512
+
+Module: `sha512`
+Digest: 512 bit (64 byte)
+
+- SHA-512/256
+
+Truncated version of of SHA-512
+
+Module: `sha512_256`
+Digest: 256 bit (32 byte)
+
+- SHA-384
+
+Truncated version of of SHA-512
+
+Module: `sha384`
+Digest: 384 bit (48 byte)
+
+### [SipHash](https://en.wikipedia.org/wiki/SipHash)
+
+SipHash is designed as a non-cryptography hash function.
+
+Module: `siphash24`
+Digest: 64 bit (8 byte)
+
+### [RIPEMD](https://en.wikipedia.org/wiki/RIPEMD)
+
+Supported functions from the RIPEMD family:
+
+- RIPEMD-160
+
+Module: `ripemd160`
+Digest: 160 bit (20 byte)
+
+### HASH160
+
+HASH-160 is defined as `RIPEMD-160(SHA-256(msg))`
+
+Module: `hash160`
+Digest: 160 bit (20 byte)
+
+### HASH256 (Double SHA256)
+
+HASH-256 is defined as `SHA-256(SHA-256(msg))`
+
+Module: `sha256d`
+Digest: 256 bit (32 byte)
+
+### Tagged SHA-256
+
+A tagged hash is defined as the `SHA-256(SHA-256(tag) || SHA-256(tag) || msg)`. In other words
+the message is prefixed with two copies of SHA-256(tag) before SHA-256 hashing the whole thing.
+
+Reference: [BIP-119](https://github.com/bitcoin/bips/blob/master/bip-0119.mediawiki)
+
+Module: `sha256t`
+Digest: 256 bit (32 byte)
+
+### [HMAC](https://en.wikipedia.org/wiki/HMAC)
+
+Hash-based Message Authentication Code (HMAC) is a specific type of [message authentication
+code](https://en.wikipedia.org/wiki/Message_authentication_code) (MAC) involving a cryptographic
+hash function and a secret cryptographic key.
+
+We implement HMAC-x where x is any of the hashes above.
```

Comment:
In ce1881bbfd8029cce25799ba0668ebae43259b9b:

We implement HMAC for any hash implementing our `Hash` trait. It is not limited to the hashes in this repository and not explicitly implemented for any of them.

Reply:
And do we want to keep it like that? A lot of the problems patching the `Hash` trait stem (in my experience) from the traits usage in `Hmac`.